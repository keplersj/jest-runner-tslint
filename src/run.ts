import * as fs from "fs";
import { Configuration, Linter } from "tslint";
import { getTslintOptions } from "./utils/normalize-config";

module.exports = ({ testPath, config }) => {
  const start = Date.now();
  const options = getTslintOptions(config);

  const fileContents = fs.readFileSync(testPath, "utf8");
  const program = Linter.createProgram("tsconfig.json", config.rootDir);

  const linter = new Linter(options, program);
  const configuration = Configuration.findConfiguration(null, testPath).results;
  linter.lint(testPath, fileContents, configuration);
  const result = linter.getResult();

  const end = Date.now();

  return {
    console: null,
    displayName: "TSLint",
    failureMessage: result.failures.length && result.output,
    numFailingTests: result.errorCount,
    numPassingTests: result.errorCount ? 0 : 1,
    numPendingTests: 0,
    perfStats: {
      end,
      start
    },
    skipped: false,
    snapshot: {
      added: 0,
      fileDeleted: false,
      matched: 0,
      unchecked: 0,
      unmatched: 0,
      updated: 0
    },
    sourceMaps: {},
    testExecError: null,
    testFilePath: testPath,
    testResults:
      result.failures.length !== 0
        ? result.failures.map(failure => ({
            ancestorTitles: [],
            duration: end - start,
            failureMessages: [failure.getFailure()],
            fullName: undefined,
            location: {
              column: failure.getStartPosition().getLineAndCharacter()
                .character,
              line: failure.getStartPosition().getLineAndCharacter().line
            },
            numPassingAsserts: 0,
            status: "failed",
            title: failure.getFailure()
          }))
        : [
            {
              ancestorTitles: [],
              duration: end - start,
              failureMessages: [],
              fullName: undefined,
              location: undefined,
              numPassingAsserts: 1,
              status: "passed",
              title: ""
            }
          ]
  };
};
