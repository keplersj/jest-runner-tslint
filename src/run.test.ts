Date.now = jest.fn(() => 1482363367071);

import * as path from "path";
import run from "./run";

describe("TSLint Jest Runner", () => {
  describe("failing fixture", () => {
    it("matches snapshot", () => {
      const result = run({
        testPath: path.join(__dirname, "__fixtures__", "bad.ts"),
        config: { rootDir: path.join(__dirname, "__fixtures__") },
      });

      expect(result).toMatchSnapshot({
        failureMessage: expect.any(String),
        perfStats: expect.any(Object),
        testFilePath: expect.any(String),
      });
    });
  });

  describe("passing fixture", () => {
    it("matches snapshot", () => {
      const result = run({
        testPath: path.join(__dirname, "__fixtures__", "good.ts"),
        config: { rootDir: path.join(__dirname, "__fixtures__") },
      });

      expect(result).toMatchSnapshot({
        perfStats: expect.any(Object),
        testFilePath: expect.any(String),
      });
    });
  });
});
