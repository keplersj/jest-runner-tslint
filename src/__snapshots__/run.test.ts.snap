// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSLint Jest Runner failing fixture matches snapshot 1`] = `
Object {
  "console": null,
  "displayName": "TSLint",
  "failureMessage": "/home/kepler/Code/keplersj/jest-runner-tslint/src/__fixtures__/bad.ts[8m:3:5[28m
[31mERROR: 3:5[39m  [90mno-console[39m  [33mCalls to 'console.log' are not allowed.[39m
",
  "leaks": false,
  "numFailingTests": 1,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": Array [],
  "skipped": false,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        "Calls to 'console.log' are not allowed.",
      ],
      "fullName": undefined,
      "location": Object {
        "column": 4,
        "line": 2,
      },
      "numPassingAsserts": 0,
      "status": "failed",
      "title": "Calls to 'console.log' are not allowed.",
    },
  ],
}
`;

exports[`TSLint Jest Runner passing fixture matches snapshot 1`] = `
Object {
  "console": null,
  "displayName": "TSLint",
  "failureMessage": 0,
  "leaks": false,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": Array [],
  "skipped": false,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [],
      "fullName": undefined,
      "location": undefined,
      "numPassingAsserts": 1,
      "status": "passed",
      "title": "",
    },
  ],
}
`;
